lib/datadog/tracing/distributed/headers/datadog.rb

```diff
@@ -1,49 +1,130 @@
+# frozen_string_literal: true
 # typed: true
 
 require_relative 'parser'
 require_relative 'ext'
 require_relative '../../trace_digest'
+require_relative '../datadog_tags_codec'
 
 module Datadog
   module Tracing
     module Distributed
       module Headers
         # Datadog provides helpers to inject or extract headers for Datadog style headers
         module Datadog
-          include Ext
+          class << self
+            include Ext
 
-          def self.inject!(digest, env)
-            return if digest.nil?
+            def inject!(digest, env)
+              return if digest.nil?
 
-            env[HTTP_HEADER_TRACE_ID] = digest.trace_id.to_s
-            env[HTTP_HEADER_PARENT_ID] = digest.span_id.to_s
-            env[HTTP_HEADER_SAMPLING_PRIORITY] = digest.trace_sampling_priority.to_s if digest.trace_sampling_priority
-            env[HTTP_HEADER_ORIGIN] = digest.trace_origin.to_s unless digest.trace_origin.nil?
+              env[HTTP_HEADER_TRACE_ID] = digest.trace_id.to_s
+              env[HTTP_HEADER_PARENT_ID] = digest.span_id.to_s
+              env[HTTP_HEADER_SAMPLING_PRIORITY] = digest.trace_sampling_priority.to_s if digest.trace_sampling_priority
+              env[HTTP_HEADER_ORIGIN] = digest.trace_origin.to_s unless digest.trace_origin.nil?
 
-            env
-          end
+              inject_tags(digest, env)
+
+              env
+            end
+
+            def extract(env)
+              # Extract values from headers
+              headers = Parser.new(env)
+              trace_id = headers.id(HTTP_HEADER_TRACE_ID)
+              parent_id = headers.id(HTTP_HEADER_PARENT_ID)
+              origin = headers.header(HTTP_HEADER_ORIGIN)
+              sampling_priority = headers.number(HTTP_HEADER_SAMPLING_PRIORITY)
+
+              # Return early if this propagation is not valid
+              # DEV: To be valid we need to have a trace id and a parent id
+              #      or when it is a synthetics trace, just the trace id.
+              # DEV: `Parser#id` will not return 0
+              return unless (trace_id && parent_id) || (origin && trace_id)
+
+              trace_distributed_tags = extract_tags(headers)
+
+              # Return new trace headers
+              TraceDigest.new(
+                span_id: parent_id,
+                trace_id: trace_id,
+                trace_origin: origin,
+                trace_sampling_priority: sampling_priority,
+                trace_distributed_tags: trace_distributed_tags,
+              )
+            end
+
+            private
+
+            # Export trace distributed tags through the `x-datadog-tags` header.
+            def inject_tags(digest, env)
+              return if digest.trace_distributed_tags.nil? || digest.trace_distributed_tags.empty?
+
+              if ::Datadog.configuration.tracing.x_datadog_tags_max_length <= 0
+                active_trace = Tracing.active_trace
```

Comment:
I don't think this is in the spirit of `TraceDigest`.

The purpose of introducing `TraceDigest` was in part to decouple propagation from active trace state. Instead, the developer is meant to work with a "bill of materials" which describes the trace rather than the trace itself. These changes represent a kind of inversion: instead of "trace --> digest --> propagation" its "digest --> propagation --> trace".

My concerns are with accessing and mutating the active trace from this module. I haven't yet totally wrapped my head around the needs of tag propagation and why this needs to be done this way, but I would like to avoid this, if possible.